js/store.js

```diff
@@ -112,51 +216,139 @@ const store = {
+  },
+
+  // check if third party is on the whitelist (owned by the first party)
+  // returns true if it is and false otherwise
+  onWhiteList(firstPartyFromRequest, thirdPartyFromRequest) {
+    if (thirdPartyFromRequest && this.firstPartyWhiteList) {
+      const hostnameVariantsFirstParty
+        = this.getHostnameVariants(firstPartyFromRequest);
+      for (let i = 0; i < hostnameVariantsFirstParty.length; i++) {
+        if (this.firstPartyWhiteList
+          .hasOwnProperty(hostnameVariantsFirstParty[i])) {
+          // first party is in the whitelist
+          const index = this.firstPartyWhiteList[hostnameVariantsFirstParty[i]];
+          const hostnameVariantsThirdParty
+            = this.getHostnameVariants(thirdPartyFromRequest);
+          for (let j = 0; j < hostnameVariantsThirdParty.length; j++) {
+            if (this.thirdPartyWhiteList[index]
+              .includes(hostnameVariantsThirdParty[j])) {
+              return true;
+            }
+          }
+          return false;
+        }
+      }
+    }
+    return false;
   },
 
   setFirstParty(hostname, data) {
     if (!hostname) {
       throw new Error('setFirstParty requires a valid hostname argument');
     }
+    const isNewWebsite = this.isNewWebsite(hostname);
+
     this.setWebsite(hostname, data);
+
+    if (isNewWebsite) {
+      this.updateChild(this.outputWebsite(hostname, this._websites[hostname]));
+    }
   },
 
+  // @todo use data from capture to update UI/filter graph
+  // eslint-disable-next-line no-unused-vars
   setThirdParty(origin, target, data) {
-    let newThirdParty = false;
+    let isNewWebsite = false;
     if (!origin) {
       throw new Error('setThirdParty requires a valid origin argument');
     }
 
-    const firstParty = this.getWebsite(origin);
-    const thirdParty = this.getWebsite(target);
+    let firstParty = this.getWebsite(origin);
+    let thirdParty = this.getWebsite(target);
 
     if (!('thirdPartyHostnames' in firstParty)) {
       firstParty['thirdPartyHostnames'] = [];
     }
     if (!firstParty['thirdPartyHostnames'].includes(target)) {
-      firstParty['thirdPartyHostnames'].push(target);
-      newThirdParty = true;
+      // third party is not currently linked to first party
+      ({ firstParty, thirdParty, isNewWebsite }
+        = this.whiteListCheck(origin, target, firstParty, thirdParty));
     }
-    if (!('firstPartyHostnames' in firstParty)) {
+    if (!('firstPartyHostnames' in thirdParty)) {
       thirdParty['firstPartyHostnames'] = [];
     }
     if (!thirdParty['firstPartyHostnames'].includes(origin)) {
       thirdParty['firstPartyHostnames'].push(origin);
-      newThirdParty = true;
     }
 
-    this.setWebsite(origin, firstParty);
+    this.setFirstParty(origin, firstParty);
     this.setWebsite(target, thirdParty);
 
-    if (newThirdParty) {
-      this.updateChild(this.outputWebsite(target, data, origin));
+    if (isNewWebsite && thirdParty['isVisible']) {
```

Comment:
I think we can simplify this further by making this one function. I don't think there is an obvious way to make this a shorter function without moving some of the initialising our into the set or get functions perhaps but for now I think this is fine:

Please test this (perhaps in a branch from this branch):

```js
setThirdParty(origin, target, data) {
  if (!origin) { 
    throw new Error('setThirdParty requires a valid origin argument');
  } 

  let isNewThirdParty = false;
  let firstParty = this.getWebsite(origin);
  let thirdParty = this.getWebsite(target);

  if (!('firstPartyHostnames' in thirdParty)) { 
    thirdParty['firstPartyHostnames'] = [];
  } 
  if (!thirdParty['firstPartyHostnames'].includes(origin)) { 
    thirdParty['firstPartyHostnames'].push(origin);
  } 

  // third party is not currently linked to first party
  if (!firstParty['thirdPartyHostnames'].includes(target)) { 
    /* When the third party isn't visible or we are not sure,
       we check the whitelist */
    if (!('isVisible' in thirdParty)
        || !thirdParty.isVisible) { 
      /* When we find a first party not on the whitelist we choose to show the third party,
         which requires sending out updates to anything listening */
      if (!this.onWhiteList(origin, target)) { 
        thirdParty['isVisible'] = true;
        isNewThirdParty = true;
        thirdParty['firstPartyHostnames'].forEach((firstPartyHostname) => { 
          this.addFirstPartyRelationship(firstPartyHostname, target);
        });
      } 
    } 
  } 

  this.setWebsite(target, thirdParty);

  if (isNewThirdParty && thirdParty['isVisible']) { 
    this.updateChild(this.outputWebsite(target, this._websites[target]));
  } 
},

addFirstPartyRelationship(firstPartyHostname, thirdPartyHostname) {
  const firstPartyInstance = this.getWebsite(firstPartyHostname);
  if (!('thirdPartyHostnames' in firstPartyInstance)) { 
    firstPartyInstance['thirdPartyHostnames'] = [];
  } 
  if (!firstParty['thirdPartyHostnames'].includes(target)) { 
    firstPartyInstance['thirdPartyHostnames'].push(target);
    this.setFirstParty(firstPartyHostname, firstPartyInstance);
  }
}
```